pkg/model/multiversion/delta.go

```diff
@@ -0,0 +1,307 @@
+// License is located at
+//
+//     http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package multiversion
+
+import (
+	"fmt"
+	"sort"
+
+	awssdkmodel "github.com/aws/aws-sdk-go/private/model/api"
+
+	ackmodel "github.com/aws-controllers-k8s/code-generator/pkg/model"
+)
+
+// FieldChangeType represent the type of field modification.
+//
+// Currently there 8 supported FieldChangeTypes
+// - FieldChangeTypeUnknown is used when ChangeType cannot be computed.
+// - FieldChangeTypeNone is used when a field name and structure didn't change.
+// - FieldChangeTypeAdded is used when a new field is introduced in a CRD.
+// - FieldChangeTypeRemoved is used a when a field is removed from a CRD.
+// - FieldChangeTypeRenamed is used when a field is renamed.
+// - FieldChangeTypeShapeChanged is used when a field shape has changed.
+// - FieldChangeTypeShapeChangedFromStringToSecret is used when a field change to
+//   a k8s secret type.
+// - FieldChangeTypeShapeChangedFromSecretToString is used when a field changed from
+//   a k8s secret to a Go string.
+type FieldChangeType string
+
+const (
+	FieldChangeTypeUnknown                        FieldChangeType = "unknown"
+	FieldChangeTypeNone                           FieldChangeType = "none"
+	FieldChangeTypeAdded                          FieldChangeType = "added"
+	FieldChangeTypeRemoved                        FieldChangeType = "removed"
+	FieldChangeTypeRenamed                        FieldChangeType = "renamed"
+	FieldChangeTypeShapeChanged                   FieldChangeType = "shape-changed"
+	FieldChangeTypeShapeChangedFromStringToSecret FieldChangeType = "shape-changed-from-string-to-secret"
+	FieldChangeTypeShapeChangedFromSecretToString FieldChangeType = "shape-changed-from-secret-to-string"
+)
+
+// FieldDelta represent the delta between the same original field in two
+// different CRD versions. If a field is removed in the Destination version
+// the Destination value will be nil. If a field is new in the Destination
+// version, the Source value will be nil.
+type FieldDelta struct {
+	ChangeType FieldChangeType
+	// Field from the destination CRD
+	Destination *ackmodel.Field
+	// Field from the source CRD
+	Source *ackmodel.Field
+}
+
+// CRDDelta stores the spec and status deltas for a custom resource.
+type CRDDelta struct {
+	SpecDeltas   []FieldDelta
+	StatusDeltas []FieldDelta
+}
+
+// ComputeCRDFieldsDeltas compares two ackmodel.CRD instances and returns the
+// spec and status fields deltas. srcCRD is the CRD of the spoke (source) version
+// and dstCRD is the CRD of the hub (destination) version.
+func ComputeCRDFieldsDeltas(srcCRD, dstCRD *ackmodel.CRD) (*CRDDelta, error) {
```

Comment:
Can we change these parameter names to just `src` and `dst` and leave off the `CRD` part? Would make it easier to read and it's obvious from the type what the parameters are.

Reply:
Agreed. Changed.